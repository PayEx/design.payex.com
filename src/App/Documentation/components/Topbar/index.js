import React, { Component } from "react";
import PrismCode from "react-prism";

import { ComponentPreview, DocContainer, Property, Attribute } from "#";
import TopbarComponent from "@/Topbar";
import { topbar } from "$/px-script/main";

const menu = {
    btn: {
        icon: "menu",
        text: "Menu"
    },
    items: ["Link 1", "Link 2", "Link 3"]
};

const Overview = () => (
    <>
        <h2 id="overview">Overview</h2>
        <p>
            To use a topbar the bare minimum markup required is:
        </p>
        <ul>
            <li><Property value=".topbar" /> wrapper.</li>
            <li>Anchor with class <Property value=".topbar-logo" /> to display the logo and make it clickable.</li>
        </ul>
        <p>
            With this you get a topbar with the PayEx logo.
            In the topbar you can add additional functionality like a menu containing links. To enable this add a <PrismCode className="language-html">{"<button>"}</PrismCode> and
            a <PrismCode className="language-html">{"<nav>"}</PrismCode> containing anchors. Remember to include <Attribute data name="toggle-nav" value="{your_nav_id}"/> as a <PrismCode className="language-html">{"<button>"}</PrismCode> attribute
            to get the menu to appear when the menu-button is clicked.
        </p>
        <ComponentPreview language="html" showCasePanel codeFigure>
            <TopbarComponent topbarContent={menu} logout />
        </ComponentPreview>
    </>
);

class Topbar extends Component {
    componentDidMount () {
        topbar.init(1);
    }

    render () {
        return (
            <DocContainer docToc>
                <p className="lead">The topbar is used to give users an easily available navigational bar at the top of your web application.</p>
                <Overview />
            </DocContainer>
        );
    }
}

export default Topbar;

/* For testing */
export { Overview };
