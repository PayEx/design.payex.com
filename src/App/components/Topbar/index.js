import React, { Fragment } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

const TopbarBtn = ({ icon, text }) => {
    const iconClass = classnames("material-icons", icon === "menu" ? "topbar-menu-icon" : "");

    if (icon || text) {
        return (
            <button type="button" className="topbar-menu-button">{"\n\t\t"}
                    <>
                        <i className={iconClass}>{icon}</i>
                        {"\n\t\t"}
                    </>
            </button>
        );
    }

    return null;
};

const TopbarMenu = ({ menu, logout }) => {
    const { items } = menu;

    return (
        <nav className="topbar-nav">
            <div className="topbar-link-container">{"\n"}
                <i className="material-icons close-topbar-nav">close</i>
                {items.map((item, i) => (
                    <Fragment key={i}>
                        {"\n"}
                        <a href="#" className={item.name === "Home" ? "active" : null} onClick={e => e.preventDefault()}>
                            {"\n"}
                            {item.icon ? <i className="material-icons">{item.icon}</i> : null}
                            {"\n"}
                            <span>{item.name}</span>
                            {"\n"}
                        </a>
                    </Fragment>
                ))}{"\n"}
                {logout ?
                    <>
                        <a className="topbar-link-right" href="#" onClick={e => e.preventDefault()}>
                            {"\n"}
                            <i className="material-icons">exit_to_app</i>
                            {"\n"}
                            <span>Log out</span>
                            {"\n"}
                        </a>
                        {"\n"}
                    </> : null}
            </div>
        </nav>
    );
};

const Topbar = ({ topbarContent, wide, logout }) => (
    <header className={`topbar${wide ? ` topbar-${wide}-wide` : ""}`}>{"\n"}
        {topbarContent ?
        <>
            <TopbarBtn icon={topbarContent.btn.icon} text={topbarContent.btn.name} target={topbarContent.id} />{"\n"}
            {"\n"}<a href="#" className="topbar-logo" onClick={e => e.preventDefault()}></a>{"\n"}
            <TopbarMenu menu={topbarContent} logout={!!logout} />
        </> :
        <>
            {"\n"} <a href="#" className="topbar-logo" onClick={e => e.preventDefault()}></a>{"\n"}
            {logout ?
                    <>
                        <a className="topbar-link-right" href="#" onClick={e => e.preventDefault()}>
                            {"\n"}
                            <i className="material-icons">exit_to_app</i>
                            {"\n"}
                            <span>Log out</span>
                            {"\n"}
                        </a>
                        {"\n"}
                    </> : null}
        </>}
    </header>
);

Topbar.propTypes = {
    topbarContent: PropTypes.object,
    fixed: PropTypes.bool,
    logout: PropTypes.bool
};

export default Topbar;
