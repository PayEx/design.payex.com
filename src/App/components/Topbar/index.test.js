import React from "react";
import { shallow, mount } from "enzyme";

import Topbar from "./index";

describe("Component: Topbar - ", () => {
    const div = document.createElement("div");

    document.body.appendChild(div);

    const menu = {
        btn: {
            name: "Menu",
            icon: "menu"
        },
        items: [
            {
                name: "Home",
                icon: "home"
            },
            {
                name: "Purchase history",
                icon: "shopping_cart"
            },
            {
                name: "Settings",
                icon: "settings"
            }
        ]
    };

    it("is defined", () => {
        expect(Topbar).toBeDefined();
    });

    it("renders a topbar with only a clickable logo", () => {
        const wrapper = shallow(<Topbar />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).toContain("topbar-logo");
        expect(wrapper.html()).not.toContain("topbar-nav");
    });

    it("renders logout link and adds topbar-link-right to the logout", () => {
        const wrapper = shallow(<Topbar logout/>);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).toContain("Log out");
        expect(wrapper.html()).not.toContain("topbar-nav");
        expect(wrapper.find(".topbar-link-right").html()).toContain("Log out");
    });

    it("renders topbar-link-container and populates it with given information without a logout link", () => {
        const wrapper = mount(<Topbar topbarContent={menu} />);

        const renderedTopbar = wrapper.find(".topbar");
        const renderedNav = renderedTopbar.find(".topbar-nav");

        expect(wrapper).toMatchSnapshot();
        expect(renderedTopbar).toBeTruthy();
        expect(renderedNav).toBeTruthy();

        menu.items.forEach(item => expect(renderedNav.html()).toContain(item.name));

        expect(renderedNav.html()).not.toContain("Log out");
    });

    it("renders topbar-link-container and a logout link", () => {
        const wrapper = mount(<Topbar topbarContent={menu} logout />);
        const renderedTopbar = wrapper.find(".topbar");
        const renderedNav = renderedTopbar.find(".topbar-nav");

        expect(wrapper).toMatchSnapshot();
        expect(renderedTopbar).toBeTruthy();
        expect(renderedNav).toBeTruthy();
        expect(renderedNav.find(".topbar-link-right .material-icons").text()).toEqual("exit_to_app");
    });

    it("renders a topbar-menu-button when content is provided", () => {
        const wrapper = mount(<Topbar topbarContent={menu} />);

        const renderedTopbar = wrapper.find(".topbar");
        const topbarMenuBtn = renderedTopbar.find(".topbar-menu-button");

        expect(wrapper).toMatchSnapshot();
        expect(renderedTopbar).toBeTruthy();
        expect(topbarMenuBtn).toBeTruthy();
    });

    it("logo prevents default when clicked", () => {
        const wrapper = shallow(<Topbar />);
        const eventHandler = { preventDefault: jest.fn() };
        const logo = wrapper.find(".topbar-logo");

        expect(wrapper).toMatchSnapshot();

        logo.simulate("click", eventHandler);

        expect(eventHandler.preventDefault).toHaveBeenCalled();
    });

    it("menu links prevents default when clicked", () => {
        const wrapper = mount(<Topbar topbarContent={menu} />);
        const eventHandler = { preventDefault: jest.fn() };
        const menuLinks = wrapper.find(".topbar-nav a");

        menuLinks.forEach(anchor => {
            anchor.simulate("click", eventHandler);
        });

        expect(wrapper).toMatchSnapshot();
        expect(eventHandler.preventDefault).toHaveBeenCalled();
    });
});
