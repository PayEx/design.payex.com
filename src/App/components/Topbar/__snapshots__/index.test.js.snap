// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Topbar -  logo prevents default when clicked 1`] = `
<header
  className="topbar"
>
  

  

   
  <a
    className="topbar-logo"
    href="#"
    onClick={[Function]}
  />
  

</header>
`;

exports[`Component: Topbar -  menu links prevents default when clicked 1`] = `
<Topbar
  topbarContent={
    Object {
      "btn": Object {
        "icon": "menu",
        "name": "Menu",
      },
      "items": Array [
        Object {
          "icon": "home",
          "name": "Home",
        },
        Object {
          "icon": "shopping_cart",
          "name": "Purchase history",
        },
        Object {
          "icon": "settings",
          "name": "Settings",
        },
      ],
    }
  }
>
  <header
    className="topbar"
  >
    

    <TopbarBtn
      icon="menu"
      text="Menu"
    >
      <button
        className="topbar-menu-button"
        type="button"
      >
        
		
        <i
          className="material-icons topbar-menu-icon"
        >
          menu
        </i>
        
		
      </button>
    </TopbarBtn>
    

    

    <a
      className="topbar-logo"
      href="#"
      onClick={[Function]}
    />
    

    <TopbarMenu
      logout={false}
      menu={
        Object {
          "btn": Object {
            "icon": "menu",
            "name": "Menu",
          },
          "items": Array [
            Object {
              "icon": "home",
              "name": "Home",
            },
            Object {
              "icon": "shopping_cart",
              "name": "Purchase history",
            },
            Object {
              "icon": "settings",
              "name": "Settings",
            },
          ],
        }
      }
    >
      <nav
        className="topbar-nav"
      >
        <div
          className="topbar-link-container"
        >
          

          <i
            className="material-icons close-topbar-nav"
          >
            close
          </i>
          

          <a
            className="active"
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              home
            </i>
            

            <span>
              Home
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              shopping_cart
            </i>
            

            <span>
              Purchase history
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              settings
            </i>
            

            <span>
              Settings
            </span>
            

          </a>
          

        </div>
      </nav>
    </TopbarMenu>
  </header>
</Topbar>
`;

exports[`Component: Topbar -  renders a topbar with only a clickable logo 1`] = `
<header
  className="topbar"
>
  

  

   
  <a
    className="topbar-logo"
    href="#"
    onClick={[Function]}
  />
  

</header>
`;

exports[`Component: Topbar -  renders a topbar-menu-button when content is provided 1`] = `
<Topbar
  topbarContent={
    Object {
      "btn": Object {
        "icon": "menu",
        "name": "Menu",
      },
      "items": Array [
        Object {
          "icon": "home",
          "name": "Home",
        },
        Object {
          "icon": "shopping_cart",
          "name": "Purchase history",
        },
        Object {
          "icon": "settings",
          "name": "Settings",
        },
      ],
    }
  }
>
  <header
    className="topbar"
  >
    

    <TopbarBtn
      icon="menu"
      text="Menu"
    >
      <button
        className="topbar-menu-button"
        type="button"
      >
        
		
        <i
          className="material-icons topbar-menu-icon"
        >
          menu
        </i>
        
		
      </button>
    </TopbarBtn>
    

    

    <a
      className="topbar-logo"
      href="#"
      onClick={[Function]}
    />
    

    <TopbarMenu
      logout={false}
      menu={
        Object {
          "btn": Object {
            "icon": "menu",
            "name": "Menu",
          },
          "items": Array [
            Object {
              "icon": "home",
              "name": "Home",
            },
            Object {
              "icon": "shopping_cart",
              "name": "Purchase history",
            },
            Object {
              "icon": "settings",
              "name": "Settings",
            },
          ],
        }
      }
    >
      <nav
        className="topbar-nav"
      >
        <div
          className="topbar-link-container"
        >
          

          <i
            className="material-icons close-topbar-nav"
          >
            close
          </i>
          

          <a
            className="active"
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              home
            </i>
            

            <span>
              Home
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              shopping_cart
            </i>
            

            <span>
              Purchase history
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              settings
            </i>
            

            <span>
              Settings
            </span>
            

          </a>
          

        </div>
      </nav>
    </TopbarMenu>
  </header>
</Topbar>
`;

exports[`Component: Topbar -  renders logout link and adds topbar-link-right to the logout 1`] = `
<header
  className="topbar"
>
  

  

   
  <a
    className="topbar-logo"
    href="#"
    onClick={[Function]}
  />
  

  <a
    className="topbar-link-right"
    href="#"
    onClick={[Function]}
  >
    

    <i
      className="material-icons"
    >
      exit_to_app
    </i>
    

    <span>
      Log out
    </span>
    

  </a>
  

</header>
`;

exports[`Component: Topbar -  renders topbar-link-container and a logout link 1`] = `
<Topbar
  logout={true}
  topbarContent={
    Object {
      "btn": Object {
        "icon": "menu",
        "name": "Menu",
      },
      "items": Array [
        Object {
          "icon": "home",
          "name": "Home",
        },
        Object {
          "icon": "shopping_cart",
          "name": "Purchase history",
        },
        Object {
          "icon": "settings",
          "name": "Settings",
        },
      ],
    }
  }
>
  <header
    className="topbar"
  >
    

    <TopbarBtn
      icon="menu"
      text="Menu"
    >
      <button
        className="topbar-menu-button"
        type="button"
      >
        
		
        <i
          className="material-icons topbar-menu-icon"
        >
          menu
        </i>
        
		
      </button>
    </TopbarBtn>
    

    

    <a
      className="topbar-logo"
      href="#"
      onClick={[Function]}
    />
    

    <TopbarMenu
      logout={true}
      menu={
        Object {
          "btn": Object {
            "icon": "menu",
            "name": "Menu",
          },
          "items": Array [
            Object {
              "icon": "home",
              "name": "Home",
            },
            Object {
              "icon": "shopping_cart",
              "name": "Purchase history",
            },
            Object {
              "icon": "settings",
              "name": "Settings",
            },
          ],
        }
      }
    >
      <nav
        className="topbar-nav"
      >
        <div
          className="topbar-link-container"
        >
          

          <i
            className="material-icons close-topbar-nav"
          >
            close
          </i>
          

          <a
            className="active"
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              home
            </i>
            

            <span>
              Home
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              shopping_cart
            </i>
            

            <span>
              Purchase history
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              settings
            </i>
            

            <span>
              Settings
            </span>
            

          </a>
          

          <a
            className="topbar-link-right"
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              exit_to_app
            </i>
            

            <span>
              Log out
            </span>
            

          </a>
          

        </div>
      </nav>
    </TopbarMenu>
  </header>
</Topbar>
`;

exports[`Component: Topbar -  renders topbar-link-container and populates it with given information without a logout link 1`] = `
<Topbar
  topbarContent={
    Object {
      "btn": Object {
        "icon": "menu",
        "name": "Menu",
      },
      "items": Array [
        Object {
          "icon": "home",
          "name": "Home",
        },
        Object {
          "icon": "shopping_cart",
          "name": "Purchase history",
        },
        Object {
          "icon": "settings",
          "name": "Settings",
        },
      ],
    }
  }
>
  <header
    className="topbar"
  >
    

    <TopbarBtn
      icon="menu"
      text="Menu"
    >
      <button
        className="topbar-menu-button"
        type="button"
      >
        
		
        <i
          className="material-icons topbar-menu-icon"
        >
          menu
        </i>
        
		
      </button>
    </TopbarBtn>
    

    

    <a
      className="topbar-logo"
      href="#"
      onClick={[Function]}
    />
    

    <TopbarMenu
      logout={false}
      menu={
        Object {
          "btn": Object {
            "icon": "menu",
            "name": "Menu",
          },
          "items": Array [
            Object {
              "icon": "home",
              "name": "Home",
            },
            Object {
              "icon": "shopping_cart",
              "name": "Purchase history",
            },
            Object {
              "icon": "settings",
              "name": "Settings",
            },
          ],
        }
      }
    >
      <nav
        className="topbar-nav"
      >
        <div
          className="topbar-link-container"
        >
          

          <i
            className="material-icons close-topbar-nav"
          >
            close
          </i>
          

          <a
            className="active"
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              home
            </i>
            

            <span>
              Home
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              shopping_cart
            </i>
            

            <span>
              Purchase history
            </span>
            

          </a>
          

          <a
            className={null}
            href="#"
            onClick={[Function]}
          >
            

            <i
              className="material-icons"
            >
              settings
            </i>
            

            <span>
              Settings
            </span>
            

          </a>
          

        </div>
      </nav>
    </TopbarMenu>
  </header>
</Topbar>
`;
